### ISSUE TITLE
`StopOnFirstFailure` should be resurrected!

### ISSUE THREAD

**Issue by AradAral** [FluentValidation/FluentValidation#1822]
### Is your feature request related to a problem? Please describe.

Hi there. Thank you for your amazing work.

This has been discussed a few times before — #1576, #1558 — but to be honest I'm not really satisfied with Jeremy's answer; and I'll present my reasoning in just a bit.

I have dozens of validators for different types in my project, and for nearly all of them, I want the following logic:
- At property level, I want the first failure to stop the validation chain; so that there is at most one validation error per property.
- At object (validator) level, I want the validation to continue for all properties, regardless of whether one or more of them fails; so that there are at most as many validation errors as there are properties in the object.

This is a very common style.

To achieve this, I either have to repeat `.Cascade(CascadeMode.Stop)` for EVERY SINGLE PROPERTY which is tedious and verbose, or I can use the deprecated `StopOnFirstFailure` either at validator class level (`CascadeMode = CascadeMode.StopOnFirstFailure`) or in the global configurations (`ValidatorOptions.Global.CascadeMode = CascadeMode.StopOnFirstFailure;`).
But of course `StopOnFirstFailure` is marked as deprecated, and as such causes an annoying warning, plus Jeremy [has advised against using it](https://github.com/FluentValidation/FluentValidation/issues/1576#:~:text=suppress%20the%20warning.-,I%20do%20recommend%20not%20using%20it%20though,-%2C%20and%20sticking%20with) and argues that it's better to repeat `.Cascade(CascadeMode.Stop)` for every single property instead.
I strongly disagree. The whole point of setting validator-level or global configurations, in general, is to not have to explicitly repeat that configuration in each specific case.

The most important thing to remember is that **what the `StopOnFirstFailure` option DOES is in fact perfectly legitimate, the only problem it ever had was its confusing NAME, which is what eventually led to it being marked as deprecated.**
But I would argue it shouldn't have been, it should've just been renamed. Because again, wanting one failure per rule chain is a perfectly legitimate and common requirement, and if you want this behavior across all your validators, as I do, ideally you should be able to specify this configuration in one place and not have to repeat it over and over again. That's the whole point of global configurations.

### Describe the solution you'd like

I'd propose the following:
Either rename `Stop` and `StopOnFirstFailure` to phrases that more accurately describe their behavior, with their behavior remaining intact.
Sure, this is a breaking change, but it's a small one anyway, and it would once and for all solve this age-old confusion over these two options. I'm confident that the people who'd be happy about this change would certainly outnumber those who'd be unhappy and would wish the old names back.

But if you absolutely don't want any breaking change:
Resurrect `StopOnFirstFailure` and remove the `[Obsolete]` attribute on top of it.
I know you can suppress the warning in your code manually, but I'm sure you'd agree that it's really ugly. More importantly, if the whole point of marking `StopOnFirstFailure` as deprecated is to make people aware of its difference with `Stop` — which it seems to be because Jeremy has stated multiple times that it's not going to be removed — then you can also do that in its summary. Marking it as deprecated is not semantically correct, and it sends the wrong message, by seeing the word "obsolete" in the tooltip when hovering over it, one would think that that means it shouldn't be used at all, or even worse, that it's going to be removed in future versions; neither of which is true in the case of `StopOnFirstFailure`. So, that's why I believe `StopOnFirstFailure` is not really "deprecated" in any meaningful sense. The only thing wrong with it is its name, which is not sufficient reason to mark it as deprecated.

### Additional Context

And once again, I'm not the first person to have been bothered by this:

[Here](https://github.com/FluentValidation/FluentValidation/issues/1718#:~:text=for%20what%20its%20worthy%2C%20I%20think%20you%20should%20rename%20it%20to%20something%20better%2C%20dont%20deprecated%20it%20because%20there%20should%20be%20a%20way%20to%20do%20it%20at%20validator%20level%2C%20just%20with%20a%20better%20name.) by @zoinkydoink:
> for what its worthy, I think you should rename it to something better, dont deprecated it because there should be a way to do it at validator level, just with a better name.

[Here](https://github.com/FluentValidation/FluentValidation/issues/1576#:~:text=We%20were%20using%20CascadeMode.StopOnFirstFailure%20in%20a%20many%20of%20our%20projects%20(at%20global%20and%20sometimes%20validator%20level)%2C%20but%20this%20has%20now%20been%20deprecated.%20We%20have%20ways%20to%20reproduce%20the%20old%20behavior%2C%20but%20finding%20it%20a%20little%20verbose%20using%20the%20new%20CascadeMode%20options.) by @dave-allen-cko:
> We were using CascadeMode.StopOnFirstFailure in a many of our projects (at global and sometimes validator level), but this has now been deprecated. We have ways to reproduce the old behavior, but finding it a little verbose using the new CascadeMode options.

[Here](https://github.com/FluentValidation/FluentValidation/issues/1500#issuecomment-704522953) by @buvinghausen:
> I think the ability to fail fast is good but typically speaking I want to give the user feedback on all the failed properties at once what I don't like is giving them multiple lines of feedback per property. As such the stop on first failure set at the validator level was exactly what I wanted. Now I have to add the cascade call to every single rule for which is tedious and verbose when you're dealing with so many properties.

etc.

**Comment by buvinghausen** [comment]
Not to answer for Mr Skinner @JeremySkinner but he never removed stop on first failure it has just been marked obsolete since I believe 9.0.  I just wrap my enum usage in a pragma warning disable and have moved on as it still works as of 10.3.3.  We could argue whether or not it should be obsoleted or not clearly but the functionality is still there.

```csharp
#pragma warning disable 618
ValidatorOptions.Global.CascadeMode = CascadeMode.StopOnFirstFailure;
#pragma warning restore 618
```

**Comment by JeremySkinner** [comment]
Hi,

> The most important thing to remember is that what the StopOnFirstFailure option DOES is in fact perfectly legitimate, the only problem it ever had was its confusing NAME, which is what eventually led to it being marked as deprecated.

The behaviour is also confusing, which is why I no longer recommend using it. 

> Either rename Stop and StopOnFirstFailure to phrases that more accurately describe their behavior, with their behavior remaining intact.

Any suggestions for what you'd prefer them to be called?
 
Personally I'm not keen on making any further changes here, but am open to suggestions.

**Comment by AradAral** [comment]
Hi @JeremySkinner. Thanks for the response.
> The behaviour is also confusing, which is why I no longer recommend using it.

Personally, I don't think the behavior is confusing at all. Because in every case, once it's explicitly explained to people, almost everyone easily understands what it DOES.
To say that _the behavior_ is confusing is to say that even after it's made clear and explained, people still don't get it, which isn't true in this case.
The root of all this confusion has always been the name, and I admit, the name is very ambiguous, as it can easily be interpreted to mean that the option does what now `Stop` does. But again, let's not forget that had the names been clearer from the beginning, probably at least 90% of this confusion wouldn't occur. Wouldn't you agree?

> Personally I disagree; I think being explicit here is much clearer. Having this as a global configuration option has caused a huge amount of confusion over the last 10 years...

Again, same point. What confused people was the apparent mismatch between the name and the underlying functionality, and not the underlying functionality per se.
The underlying functionality requires no more than 50 words to clearly explain, and almost everyone would understand it after reading that explanation.
In addition to this, please also consider the fact that before `StopOnFirstFailure` was made deprecated, probably many people used it WITHOUT being confused, but we didn't see them here because they didn't create GitHub issues, the only ones that did create GitHub issues, expressing their confusion over this, were those that were confused! There could've been as many or more people who were NOT confused. But anyways, that's a side point. All I'm trying to say is that the underlying functionality is valid, the only thing that was ever wrong with this option, was its name.

> Any suggestions for what you'd prefer them to be called?

Here's my suggestion:
- `OneFailurePerValidator`:
When the first rule of the first rule chain of the validator fails, the execution stops altogether, and the rest of the rule chains will not be executed; such that at most there's only one reported failure per validator. Previously called `Stop`.

- `OneFailurePerRuleChain`:
When the first rule of a rule chain fails, the execution within that rule chain stops and moves to the next rule chain; such that at most there's one reported failure per rule chain. Previously called `StopOnFirstFailure`.

- `Continue`:
The execution doesn't stop at any failure, so all rules across all rule chains are executed; such that at most there are as many reported failures as there are rules in the validator.

This is just one suggestion off the top of my head, if you're open to the renaming of these options then we can think about alternative names more, and the descriptions can also be tweaked to your liking. But I would like to know that apart from a small breaking change, what downside do you think there is to renaming these two and thereby settling this age-old conversation once and for all. :D
Let me know what you think.

**Comment by JeremySkinner** [comment]
> Personally, I don't think the behavior is confusing at all. 

So I disagree with this, having seen the questions that it's raised even once the naming has been explained. You might have a slightly different perspective if you'd spent a long time having to support it

> please also consider the fact that before StopOnFirstFailure was made deprecated, probably many people used it WITHOUT being confused

I've seen a lot of use of the library "in the wild" too, and seen it being used incorrectly a lot of the time...my comments aren't just 
based on opened github issues (but you're right though, opened issues will obviously have a bias towards people who have misunderstood it), but I'd request that you please trust and accept my experience on this, thanks :) 

> What I really want to know is that apart from a small breaking change, what downside do you think there is

So the issue is that although it appears a small breaking change on paper, in reality this is the kind of thing that affects a lot of people, which in turn generates more support requests, which I actively want to avoid. 

That being said I am still very open to renaming and reintroducing. I personally don't have the time or the desire to work on this feature anymore, but if you'd like to pick it up and take on development for this then I'd be happy to provide guidance. 

Some initial questions that come to mind:
- Would `OneFailurePerValidator` be better named `OneFailurePerClass`, as there may be some confusion as to what "per validator" means (eg could it be misunderstood as "one failure per property validator" rather than "one failure per validator class")
- What should happen if you set `OneFailurePerValidator` inside a rule chain: `RuleFor(x => x.Foo).Cascade(CascadeMode.OneFailurePerValidator)`. Technically, it'd behave the same as if you used `OneFailurePerRuleChain` here, but from a readability point of view this is confusing and doesn't really make sense. Perhaps rule-level cascade should use a different enum that doesn't have this option?
- I think we should preserve the existing Stop/StopOnFirstFailure options at least for 1 major version, so the new options would need to be additions rather than straight renames.

**Comment by jsabrooke** [comment]
Sorry to continue this one @JeremySkinner , I'm sure that's what you wanted :-). I keep forgetting how these cascade modes work and it's the third time I've ended up looking at this thread, so thought I'd finally see if I can contribute!

I have to say I totally agree with basically all of @AradAral's reasons for wanting the old behavior working again (without a warning), specifically that the functionality is very reasonable, and I think that it was _mainly_ the naming that brought about the confusion originally. Your responses also make total sense @JeremySkinner, it's an awkward one! Particularly the questions you raised about @AradAral's suggestion -

You're clearly right that having the ability to set `CascadeMode.OneFailurePerValidator` inside a rule chain wouldn't make sense, and shouldn't be possible. In fact it's similar to an issue present now - you can set `Stop` or `StopOnFirstFailure` inside a rule chain, but both will do the same thing, unless I'm mistaken?

I think the solution could be exactly what you said though @JeremySkinner - have two enums. This would enable all the useful functionality without any warnings, and also be totally clear (once the deprecated bits eventually get removed; in the meantime it would be _pretty_ clear I think :-) ). Also without any subtle breaking changes. None initially, and then obvious and easy-to-fix compiler errors when deprecated bits eventually removed.

Details of my suggestion below under the big headings. What do you think? I suppose I'd be willing to do the development for it if you think it's reasonable, or if you have any suggested changes? 

An alternative I considered would be to have only one enum with `Continue` and `Stop` (and `StopOnFirstFailure (deprecated)` that gets removed eventually), but different properties or methods on the validator. Something like `Validator.SetDefaultRuleChainCascadeMode(CascadeMode)` and `Validator.SetDefaultCascadeMode(CascadeMode)`. Or something. Anyway I decided the extra enum would probably be easier so gave up on that.

# Existing Functionality
Just to be complete, the existing functionality is this (please let me know if I'm wrong):
	
### At validator-level:
  `Continue`: Continues execution of the validator on any failure.
  `Stop`: Stops execution of the validator on any failure at any time.
  `StopOnFirstFailure (deprecated)`: For each rule chain in the validator, stop execution of rule chain on any failure, but then continue to the next rule chain. Same as setting Stop on all rule chains individually.

### At rule-chain-level
Setting here overrides validator-level setting for the specific rule chain
  `Continue`: Continues execution of the rule chain on any failure.
  `Stop`: Stops execution of the rule chain on any failure.
  `StopOnFirstFailure (deprecated)`: Same as `Stop`.

### At global-level
Sets the default value at validator-level.

# My suggestion

### At validator-level
Use the existing `CascadeMode` enum but adjusted as follows:
	`Continue`: Continues execution of the validator on any failure.
	`Stop (deprecated)`: Renamed to `StopOnAnyFailure`
	`StopOnFirstFailure (deprecated)`: Renamed to `StopRuleChainOnFailure`
	`StopOnAnyFailure`: Stops execution of the validator on any failure at any time  [Existing `Stop` functionality].
	`StopRuleChainOnFailure` For each rule chain in the validator, stop on any failure, but continue to the next rule chain. Same as setting Stop on all rule chains individually [Existing `StopOnFirstFailure` functionality].

### At rule-chain-level
Use a new enum called `RuleChainCascadeMode`.
	`Continue`: Continues execution of the rule chain on any failure.
	`Stop`: Stops execution of the rule chain on any failure.

Add an overload of .Cascade in a rule chain to accept `RuleChainCascadeMode`, and change wherever the passed value is stored (presumably a field/property somewhere) from `CascadeMode` to the new `RuleChainCascadeMode`. Mark the existing rule chain `Cascade(CascadeMode...` method as obsolete, and change its behaviour so that it just sets the new value - passing `CascadeMode.Continue` sets it to `RuleChainCascadeMode.Continue`, and passing any of the other four sets it to `CascadeMode.Stop`.

Then get rid of all the deprecated bits whenever you're willing to, and make it clear in the main docs how to fix the new errors (would be pretty simple conversion for everyone).

> I do wonder though if it's worth the effort vs just letting people continue to use StopOnFirstFailure with a warning suppression if they really want to...

Personally, I really think it is :-).

**Comment by JeremySkinner** [comment]
Hi @jsabrooke thanks for your perspective.

> Personally, I really think it is :-).

From the maintainer's perspective these are the things that currently make it not worth the effort:

- This isn't a feature I personally have any interest in 
- I wouldn't have the time to work on it for the foreseeable future
- As a breaking change, it'll need suitable documentation and upgrade guide which again I don't currently have time for
- There will inevitably be a bunch of support requests (as there always are for breaking changes)
- The old functionality is not commonly used from what I've seen of FV's use "in the wild"

vs keeping things as they are:

- Doesn't require any time investment from me
- Users such as yourself that want to continue to use the old behaviour can still do so, although you'll have to suppress the warning

So for me personally this isn't worth it.

However, if this is something that you (or others in the community) _strongly_ feel should be in the library then I'm open to it, but it'll need someone to take ownership of this, both in terms of the development and ongoing support as I don't have the capacity. So practically this would mean:

- Putting together a PR targeting 11.x that adds the new functionality, the necessary backwards compatibility overloads, and appropriate tests (I'd be available for any guidance/questions & review) 
- Updates to the documentation
- New section(s) in the 11.x Upgrade Guide
- A willingness and commitment to provide support (and if necessary fixes) for any issues raised related to the new functionality on this issue tracker going forward. 

Let me know what you think.

**Comment by jsabrooke** [comment]
Hi @JeremySkinner, thanks for our response. Sorry I took a while to get back to you; I wanted to make sure I had time to think about it first.

Yes, I would be willing to do all those things that you've said. How would providing support work, would it just be monitoring issues raised here?

If you're still ok with me doing something on this, I had some more thoughts. The only problem I can see with my previous suggestion is that there's technically one more scenario that it wouldn't cover. If you wanted to it to stop on failure at validator-level, but _not_ by default at rule-level; it wouldn't let you do that. I've got no idea if anyone would want this, but I don't like the idea of making it difficult; it would be similar to the current issue but the other way round - you'd have to set the validator to `Stop[OnAnyFailure]` and each rule individually to `Continue`.

So these are the options I can see:
1. Ignore the extra thought above and do my original suggestion (possibly removing the word "Chain" everywhere I added it, as I don't think it's needed?).
2. Stick with the extra enum suggestion, but add a third `Stop`-type value of `StopAfterRuleChainFailure`, to the others - `StopOnAnyFailure` and `StopRuleChainOnFailure` (ignoring the deprecated ones that would remain for a bit).
3. Go back to the other thought I had but discarded previously, of having one enum exactly as it is now (but eventually removing `StopOnFirstFailure`), and add a new property on `AbsractValidator` of `DefaultRuleCascadeMode`. This is then passed to new rules instead of the current `AbsractValidator.CascadeMode`. The latter now just controls cascade mode in between rules, i.e. at validator-level. So for the situation we've encountered, we would do:
```
CascadeMode = CascadeMode.Continue;
DefaultRuleCascadeMode= CascadeMode.Stop;
```
There would also need to be a similar new `DefaultRuleCascadeMode` property at global-level, so in reality we'd probably usually use that unless we want to override it in a specific validator. It might also make sense to rename `AbsractValidator.CascadeMode` to something validator-level specific, but it's hard to think of a good one. I'd say `ValidatorCascadeMode`, but then there's your earlier question of "Would OneFailurePerValidator be better named OneFailurePerClass, as there may be some confusion as to what 'per validator' means?".

4. You tell me to go away and leave you alone ;-).

My preferred option is now option 3, as I think it's clearer and is a better representation of the actual functionality. But it is also more of a change from the current code, and would need a bit more work and thought to avoid breaking changes.

Let me know your preferred option, or variation of. And fair enough if it's number 4!

**Comment by JeremySkinner** [comment]
Hi @jsabrooke 

> Yes, I would be willing to do all those things that you've said.

Great! Thanks for wanting to be involved.

> How would providing support work, would it just be monitoring issues raised here?

Yes, although realistically it's likely that I'll be monitoring issues and then tag you into anything related to this that needs your input, so you wouldn't have to monitor all the time (although of course you can if you want!). I don't allow general support/usage requests on here so it'd likely mainly be bug reports, and general questions would be covered by having good docs.

> My preferred option is now option 3

Yes I think i like option 3 best too. Happy for you to prototype this if you feel it's the best approach. 

We might need to play around with the property names a bit. Off the top of my head I think I'd prefer `ClassLevelCascadeMode` and `RuleLevelCascadeMode` on AbstractValidator. 

Anyone using current class-level `CascadeMode` property would then just pass-through to either of the new properties as appropriate (so `CascadeMode = CascadeMode.Stop` would be translated to `ClassLevelCascadeMode = CascadeMode.Stop` and `CascadeMode = CascadeMode.StopOnFirstFailure` would be translated to `RuleLevelCascadeMode = CascadeMode.Stop`).

**Comment by AradAral** [comment]
I prefer option 3 too, by the way.
Just me two cents :)
> Off the top of my head I think I'd prefer ClassLevelCascadeMode and RuleLevelCascadeMode on AbstractValidator.

I also agree that renaming the current property `CascadeMode` to something more accurate like `ClassLevelCascadeMode` makes more sense.

Also thank you @jsabrooke for taking care of this!

**Comment by AradAral** [comment]
![image](https://user-images.githubusercontent.com/26527405/149436324-7c517c1e-e4c8-4b44-ab91-e9b778ddc8d5.png)
I laughed at this harder than I should have!

**Comment by jsabrooke** [comment]
Thanks @JeremySkinner, I'll start on the code at some point soon. And sure, I'm happy with your naming suggestions etc.

And thanks @AradAral, I was hoping you'd agree with the changes actually seeing as you raised the issue!

> This made me laugh way harder than it should have! 😂

I just thought it would be a perfectly reasonable option for Jeremy to take :-)

**Comment by JeremySkinner** [comment]
This is now merged and will be in the 11.0 release

**Comment by AradAral** [comment]
I'd like to thank both @JeremySkinner for his openness and willingness to actually consider this despite being initially hesitant, I found his approach to the conversation to be very productive, and of course @jsabrooke for their absolutely great work on this.

Thank you!
